src/data-pipeline/lambda/emr-serverless-app/index.ts

```
@@ -0,0 +1,157 @@
+  logger.info(JSON.stringify(event));
+  try {
+    const res = await _handler(event, context);
+    logger.info('lambda returned', { res });
+    return res;
+  } catch (e: any) {
+    logger.error(e);
+    throw e;
+  }
+};
+
+async function _handler(event: CloudFormationCustomResourceEvent, context: Context) {
+  logger.info(`functionName: ${context.functionName}`);
+  const props = event.ResourceProperties as ResourcePropertiesType;
+  if (event.RequestType == 'Delete') {
+    await deleteEMRServerlessApp();
+    return;
+  } else {
+    const applicationId = await createEMRServerlessApp(props);
+    return {
+      Data: {
+        ApplicationId: applicationId,
+      },
+    };
+  }
+}
+
+async function createEMRServerlessApp(props: ResourcePropertiesType): Promise<string> {
+  const input: CreateApplicationCommandInput = {
+    name: props.name,
+    releaseLabel: props.version,
+    type: 'SPARK',
+    architecture: Architecture.X86_64,
+    networkConfiguration: {
+      subnetIds: props.subnetIds.split(','),
+      securityGroupIds: [props.secourityGroupId],
+    },
+    autoStartConfiguration: {
+      enabled: true,
+    },
+    autoStopConfiguration: {
+      enabled: true,
+      idleTimeoutMinutes: parseInt(props.idleTimeoutMinutes),
+    },
+  };
+
+  logger.info('CreateApplicationCommand input', { input });
+  const command = new CreateApplicationCommand(input);
+  const response = await emrServerlessClient.send(command);
+  const applicationId = response.applicationId!;
+  logger.info('created emr application applicationId:' + applicationId);
+
+  logger.info('s3Bucket:' + s3Bucket + ', appIdKey:' + appIdKey);
+  let appIdCofing = await readS3ObjectAsJson(s3Bucket, appIdKey);
+  const nowStr = process.env.TEST_TIME_NOW_STR || new Date().toISOString();
+
+  if (appIdCofing) {
+    logger.info('find appIdCofing', { appIdCofing });
+    // only save previous 5 appIds
```

Were those apps failed to be deleted?

---

if there are jobs are running, the app will be failed to delete.  if the deletion failed, just ignore the error.  the app will be deleted in next upgrade/update as we saved the previous 5 apps.